File: src/main/java/com/partymakers/shareparty/domain/party/usecase/impl/InviteFriendImpl.java
package com.partymakers.shareparty.domain.party.usecase.impl;

import com.partymakers.shareparty.domain.friends.port.FriendsRepository;
import com.partymakers.shareparty.domain.party.port.PartyRoomRepository;
import com.partymakers.shareparty.domain.party.usecase.InviteFriend;
import com.partymakers.shareparty.domain.party.usecase.exception.AlreadyExistException;
import com.partymakers.shareparty.domain.party.usecase.exception.NotFoundException;

import lombok.RequiredArgsConstructor;

@RequiredArgsConstructor
public class InviteFriendImpl implements InviteFriend {

    private final PartyRoomRepository partyRoomRepository;
    private final FriendsRepository   friendsRepository;

    @Override
    public void inviteFriend(String nickName, long partyId) {
        friendsRepository.findOneById(nickName).ifPresentOrElse(friend ->{
                partyRoomRepository.findById(partyId).ifPresentOrElse(room -> {

                        if(room.getFriends().contains(friend)){
                            throw new AlreadyExistException("Friend "  + friend.getName() + " already invited");
                        }

                        room.getFriends().add(friend);

                        partyRoomRepository.save(room);},
                    () -> {throw new NotFoundException();});
            },
        () -> {throw new NotFoundException();});
    }
}

File: src/main/kotlin/com/partymakers/shareparty/domain/party/usecase/impl/InviteFriendImpl.kt
package com.partymakers.shareparty.domain.party.usecase.impl

import com.partymakers.shareparty.domain.friends.port.FriendsRepository
import com.partymakers.shareparty.domain.party.port.PartyRoomRepository
import com.partymakers.shareparty.domain.party.usecase.InviteFriend
import com.partymakers.shareparty.domain.party.usecase.exception.AlreadyExistException
import com.partymakers.shareparty.domain.party.usecase.exception.NotFoundException

class InviteFriendImpl(
    private val partyRoomRepository: PartyRoomRepository,
    private val friendsRepository: FriendsRepository,
) : InviteFriend {

    override fun inviteFriend(nickName: String, partyId: Long) {
        val friend = friendsRepository.findOneById(nickName)
            .orElseThrow { NotFoundException() }

        val room = partyRoomRepository.findById(partyId)
            .orElseThrow { NotFoundException() }

        if (room.friends.contains(friend)) {
            throw AlreadyExistException("Friend ${friend.nickName} already invited")
        }

        val changedRoom = room.copy(friends = room.friends + friend)
        partyRoomRepository.save(changedRoom)
    }
} 